### PR TITLE
Unified app model routing

### LINKED ISSUE #47668: Unified app model routing
- add `Router` component into `App.razor`
- make `MapRazorComponent` use `App.razor`
- change `MapRazorComponens` to always render the root component (`App.razor`)
- create `RoutingStateProvider` to hold the `RouteData`
- use `RoutingStateProvider` in `Router` to avoid routing twice

### PR DESCRIPTION
**Pull request by surayya-MS**
# Unified app model routing

API Proposal https://github.com/dotnet/aspnetcore/issues/47854
Fixes #47668

### CODE REVIEW

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs
Comment (javiercn) [inline src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs:59]: I think we should also add the root component to the endpoint metadata, probably with a new `RootComponentMetadata`.

File: src/Components/Components/src/Routing/RoutingStateProvider.cs
Comment (javiercn) [inline src/Components/Components/src/Routing/RoutingStateProvider.cs:14]: Is there a scenario why someone else (other than the framework) would need to set this value?

Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:14]: Most likely no

Comment (javiercn) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:14]: If that's the case, should this be settable from outside then?

Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:14]: Yes, because I wouldn't be able to set `RouteData` in `EndpointHtmlRenderer.InitializeStandardComponentServicesAsync`.

Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:14]: I don't think there are alternatives. `RoutingStateProvider.RouteData` is set in `Components.Endpoints.EndpointHtmlRenderer` and then I need to get the value in `Components.Router` component.

Comment (javiercn) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:14]: > I don't think there are alternatives. RoutingStateProvider.RouteData is set in Components.Endpoints.EndpointHtmlRenderer and then I need to get the value in Components.Router component.

I think there are, for example, how does this work in `NavigationManager`?

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (javiercn) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:145]: Why are we filtering for null here?

In fact, why do we even need to make a copy of the dictionary at all?

Comment (surayya-MS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:145]: It is because of `RouteValueDictionay : IReadOnlyDictionary<string, object?>` and `RouteData.RouteValues` is `IReadOnlyDictionary<string, object>`. There is [CS8620](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings?f1url=%3FappId%3Droslyn%26k%3Dk(CS8620)). Another alternative is to suppress it. What do you think?

Comment (javiercn) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:145]: @surayya-MS I think we should update the nullability annotation in `RouteData.RouteValues` to match the one in `RouteValueDictionary`.

I believe it is considered a "breaking change", but that we have always sided on "correctness" in this case.

**Review by javiercn — APPROVED** [review]

File: src/Components/Components/src/Routing/RoutingStateProvider.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Routing/RoutingStateProvider.cs:15]: This approach of `RouteData` being a property is very specialized to the `EndpointHtmlRendering` scenario, since it relies on the `RouteData` value not changing over time in reaction to navigation in advance of the router getting the value. (I know a custom getter could construct the value dynamically, but that wouldn't be idiomatic)

However I think we could quite easily make this into something more general-purpose. If it was changed to:

```cs
public interface IRouteDataProvider
{
    RouteData? GetRouteData();
}
```

... then the Endpoints pipeline could register a version that behaves equivalently to what we already have here, but also people doing interactive rendering could optionally make a version to does arbitrary route URL parsing and selecting page components. That would automatically cover some other feature requests:

 * https://github.com/dotnet/aspnetcore/issues/45396
 * https://github.com/dotnet/aspnetcore/issues/45653 (maybe, didn't check super carefully)
 * https://github.com/dotnet/aspnetcore/issues/42431
 * https://github.com/dotnet/aspnetcore/issues/40736

Given how many different requests we've had for customizing the URL matching, this seems like a big bonus to get for a small design update!

File: src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs
Comment (SteveSandersonMS) [inline src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:42]: This is cool! Does it mean we can delete `RazorComponentEndpointHost` entirely now? Or is it still required in some other case?

Comment (surayya-MS) [reply src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:42]: I'm not sure. @javiercn?

File: src/Components/Components/src/Routing/RoutingStateProvider.cs
Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:15]: I didn't think about this scenario. Thanks @SteveSandersonMS ! I'll change it.

File: src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs
Comment (SteveSandersonMS) [reply src/Components/Endpoints/src/RazorComponentEndpointInvoker.cs:42]: Oh right yes, that sounds correct.

File: src/Components/Components/src/Routing/RoutingStateProvider.cs
Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:15]: I started implementing the design with `RouteData? GetRouteData(string url);` and realized that it is not quite the same as the previous design as it may seem.
Previous `RoutingStateProvider` is a scoped service that provides `RouteData` with the component type and the route parameters. Example: there is `Counter.razor` page with route "/counter/{param1}/{param2}". if you go to "/counter/test1/test2" `RoutingStateProvide` will provide `RouteData` `PageType = typeof(Counter)` and `RouteValues = {{"param1", "test1"}, {"param2", "test2"}}`. The purpuse of `RoutingStateProvider` is to provide the `RouteData` to the `Router` component so the routing wouldn't happen twice as it already happened because of endpoints.
`IRouteDataProvider` in my understanding is a singleton service that is needed for mapping routes to component type in order to override the existing routes of the pages. (and also because of that the name should be different)
Hence we cannot use `IRouteStateProvider` instead of `RoutingStateProvider` as they play different roles. If we want add a feature to map routes to a page which route is different then it is a seperate thing and we still need `RoutingStateProvider`.

Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:15]: > IRouteDataProvider in my understanding is a singleton service

I was expecting it to be a scoped service, and to do exactly what you've said that `RoutingStateProvider` would do, i.e.:

 * The default implementation would either:
   *  Just return `null` in all cases, if we're leaving `Router` to do the default `RouteData` creation
   *  ... or would contain the routedata-creation logic that currently lives in `Router`, if we're moving it into the service
 * The implementation used by `EndpointHtmlRenderer` would get routedata from what was already determined by ASP.NET Core's server-side routing system

Technically, `GetRouteData` would not need to take a `url` parameter, since implementations could depend on `NavigationManager` and get the URL from there. I only suggested having the `url` parameter since it would be easier for people who are trying to implement something like "if the URL is `/{lang}/{path}`, strip off `/{lang}` for localization and then call the default implementation passing `/{path}`".

Does this make sense? Sorry if I'm still missing something about why `IRouteDataProvider` couldn't have a role like this!

Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:15]: We just discussed this, and you clarified what was missing from my thinking. You are right that:

 * It doesn't really make sense for a centralized service to do the `url`->`RouteData` mapping on its own without having access to the table of routes declared on all the `@page` directives everywhere. Because how would the central service know how each page wants to extract parameters from the route? That information only exists in the table of routes.
   * So, people who implement `IRouteDataProvider` would almost certainly want to fall back on the default routing behavior that consults the table of routes.
 * But it **does** make sense for people to want to do URL rewriting, i.e., to perform some translation on the URL string before it goes into the default `url`->`RouteData` mapping based on the table of routes.

Looking back at the list of issues I cited above:

 * https://github.com/dotnet/aspnetcore/issues/45396: it would be sufficient to support URL rewriting
 * https://github.com/dotnet/aspnetcore/issues/45653: it would be sufficient to support URL rewriting
 * https://github.com/dotnet/aspnetcore/issues/42431: no, this is unrelated - it's about mutating the global route table somehow
 * https://github.com/dotnet/aspnetcore/issues/40736: it would be sufficient to support URL rewriting (and in fact this is an identical request to the second item in this list)

So I think there are a few options here:

 * Forget about this extra feature - just have some `RouteDataProvider` that takes no parameters and is just a way for `Endpoints` to override what `Router` does.
   * We could separately implement some kind of URL rewriting feature in the future
 * Or, move the default `Router`'s route selection logic into a new `DefaultRouteDataProvider` that has a method like `public virtual RouteData GetRouteData(string url)` so people can subclass it as a way to do URL rewriting *and* we can have a version of it in Endpoints that ignores `url` and returns whatever `RouteData` was determined by the ASP.NET Core router.

Either approach would be fine in my opinion. Perhaps you could judge whether doing the second of these options would be time-consuming or not!

Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs]: I tried to move the routing logic into the `DefaultRouteDataProvider`. It is possible if I add `RouteTable` to the API in order to trigger route table refresh from the `Router`. 
This doesn't look good because 

- If someone uses custom class that inherits `DefaultRouteDataProvider` in Blazor United application then routing is happening twice. We could make `EndpointRouteDataProvider` public and tell people that they need to inherit from `EndpointRouteDataProvider` in that case. But that is very bad.
- Also `EndpointRouteDataProvider` now has `RouteTable` property that is never used as well as in `GetRouteData` the parameter `string url` is also never used. 

I suggest we leave `RoutingStateProvider` as it was only for the purpose to avoid routing twice for Blazor United. Then in a separate PR create `DefaultRouteDataProvider` that incapsulates the routing logic of `Router` as a solution for the issues that were listed. It will use `RoutingStateProvider` inside to avoid routing twice. @javiercn also mentioned that we could add other things to `RoutingStateProvider` in the future so I think this generic name is good.

Comment (surayya-MS) [reply src/Components/Components/src/Routing/RoutingStateProvider.cs:15]: I tried to move the routing logic into `DefaultRouteDataProvider`. It is possible if I add `RouteTable` to the API to refresh route table from `Router`. This overall looks bad because

- If someone uses custom class that inherits `DefaultRouteDataProvider` in Blazor United application then routing is happening twice. We could make `EndpointRouteDataProvider` public and tell people to inherit from `EndpointRouteDataProvider` in that case but this is bad.
- There is `RouteTable` that is never used in `EndpointRouteDataProvider` as well as in `GetRouteData` parameter `string url` is also not used.

I suggest we use `RoutingStateProvider` only for the purpose to avoid routing twice in `Router`. I will revisit the issues listed previously and try to come up with another solution.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (BrennanConroy) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:87]: This cast seems unsafe since `RoutingStateProvider` is public and you're using `TryAddScoped` which implies it could be overwritten by the user. I think even with `AddScoped` it could still be replaced, but don't quote me on that.

Comment (surayya-MS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs]: Thanks @halter73! I registered `EndpointRoutingStateProvider` directly now.

> But if we're going that route (no pun intended), why not use an IRoutingStateProvider interface over a base class?

We certainly can do that. The downside to using `IRoutingStateProvider` over the base class `RoutingStateProvider` would be that in all the other places we would have same implementation of IRoutingStateProvider which is `RouteData` getter returns null. What do we usually do in these types of situations?

**Comment by surayya-MS** [comment]
Just want to clarify things about fixing `FormHandlingTest`. Because of my change now we render the root component that contains `Router` component. When I tried to use a default layout in `Router` [FormsLayout](https://github.com/dotnet/aspnetcore/blob/main/src/Components/test/testassets/Components.TestServer/RazorComponents/Shared/FormsLayout.razor) there was an exception because `CascadingModelBinder` was used twice. Once in [RouteView](https://github.com/dotnet/aspnetcore/blob/3b091f77be96f1c8555e6aa266ad4c4322af76c3/src/Components/Components/src/RouteView.cs#L79) and second in `FormsLayout`. So I just deleted `FormsLayout`. There are 2 tests that specifically don't use `CascadingModelBinder` so the components couldn't be rendered by the `Router`. For those 2 tests I created another 2 test classes with sturtups that call `MapRazorComponents<{App component with no Router and only the components that is tested}>`

File: src/Components/Samples/BlazorUnitedApp/App.razor
Comment (SteveSandersonMS) [inline src/Components/Samples/BlazorUnitedApp/App.razor:42]: With the updated project structure, the split between `App.razor` and `_Layout.razor` seems mostly arbitrary, as any of the content could go in either of them. We could even eliminate `_Layout.razor` entirely if we wanted, putting all its content into `App.razor`.

There could be some practical benefits to moving as much of the content as possible into `_Layout.razor`, though. If `App.razor` was *just* the router, then `_Layout.razor` could contain the entire HTML structure from `<html>` to `</html>`. That would have some benefits:

 * Just one place to look to see the entire common markup - no mental effort to stitch together `App.razor` and `_Layout.razor` to imagine what the full structure looks like
 * People who wanted entirely different HTML in different layouts could easily do so. For example, they could have different pages reference different CSS files in the header, etc.

I haven't checked to verify for sure this fully works end-to-end, and there's no need to block this PR on it. I'm just commenting so that we consider this when we implement the real unified project template. cc @javiercn.

File: src/Components/test/E2ETest/ServerRenderingTests/FormHandlingTests/NamedFormNoParentBindingContextTest.cs
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/ServerRenderingTests/FormHandlingTests/NamedFormNoParentBindingContextTest.cs:55]: Is this test still relevant, or could the commented-out code be removed?

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor
Comment (SteveSandersonMS) [inline src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor:30]: Could you clarify what all the `OperationId` parameters are used for? I see the parameters are declared but they don't appear to be set or used anywhere.

File: src/Components/test/E2ETest/ServerRenderingTests/FormHandlingTests/NamedFormNoParentBindingContextTest.cs
Comment (surayya-MS) [reply src/Components/test/E2ETest/ServerRenderingTests/FormHandlingTests/NamedFormNoParentBindingContextTest.cs:55]: Sorry, forgot to remove it. This test is in `FormHandlingTests/FormWithoutBindingContextTest.cs`

**Comment by SteveSandersonMS** [comment]
> Just want to clarify things about fixing FormHandlingTest

What you've done sounds reasonable to me. If this is not what @javiercn expected then perhaps he can comment when he's back.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs]: We have quite a few internal service implementations that are just no-ops because they are not relevant in a particular hosting environment. There wouldn't be anything bad about having a `WebAssemblyRoutingStateProvider` that just always returns `null` with a comment saying it's not relevant in this environment.

Alternatively, you could change `Router` so that it no longer has an `[Inject]` for the routing state provider, but instead gets it using `serviceProvider.GetService<T>()`, so that it's an optional dependency. Environments that don't supply an implementation of this service will just give `null`. Then it's only necessary to register an implementation in the endpoints case.

Comment (surayya-MS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs]: Thanks @SteveSandersonMS ! I'd better do that then.

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor
Comment (SteveSandersonMS) [reply src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor:30]: If it's not used, does it work to remove that segment from the `@page` definition? I'm not sure why it would be there if it's not used.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (surayya-MS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs]: I tried the second option and it didn't work. The reason is `serviceProvider.GetService<T>()` doesn't return null if the service wasn't registered. It throws InvalidOperationException.

If I go with the first option then I'll need new API review. I think I'm going to wait for @javiercn to come back before doing that change since he has already approved the API with class `RoutingStateProvider`.

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor
Comment (surayya-MS) [reply src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor:30]: It doesn't work if I remove it. I'll try to figure out where exactly it is used and include it.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs]: > The reason is serviceProvider.GetService<T>() doesn't return null if the service wasn't registered. It throws InvalidOperationException.

Are you sure? `GetService` is explicitly documented as returning `null` in this case: https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.serviceproviderserviceextensions.getservice?view=dotnet-plat-ext-8.0#returns

There is a separate API, `GetRequiredService`, for cases where it should throw when no implementation was registered.

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor
Comment (SteveSandersonMS) [reply src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Forms/ModifyHttpContextForm.razor:30]: Did you also update `FormsHandlingTest` to update the URLs it uses?

**Review by SteveSandersonMS — APPROVED** [review]
This looks great. Thanks for the various updates relating to feedback!

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (surayya-MS) [reply src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs]: Since I don't need `RoutingStateProvider` that returns null any more, I'm going to use interface `IRoutingStateProvider`.

File: src/Components/Samples/BlazorUnitedApp/App.razor
Comment (javiercn) [reply src/Components/Samples/BlazorUnitedApp/App.razor:42]: I will take this into consideration when I start building the template.

**Review by javiercn — APPROVED** [review]

File: src/Components/Endpoints/src/Builder/RootComponentMetadata.cs
Comment (halter73) [reply src/Components/Endpoints/src/Builder/RootComponentMetadata.cs:9]: What do we think about the `Microsoft.AspNetCore.Components.Endpoints.Builder` namespace? Do we expect much user code to reference it? If not, it's probably better than `Microsoft.AspNetCore.Builder` namespace `ComponentTypeMetadata` is currently in. But I don't think anything else exists in this namespace yet. What about [Microsoft.AspNetCore.Components.Endpoints](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.endpoints?view=aspnetcore-8.0)? I don't think this deserves its own namespace.

Stuff like the namespace can easily get missed if we try to cram everything in one big API review. Hopefully reviewing this will be fairly quick.

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs
Comment (halter73) [inline src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs:43]: Who's going to use the `RootComponentMetadata`? It doesn't look like anything references it currently.

I see that the `rootComponent` `Type` now flows to `RazorComponentEndpointInvoker` via a lambda capture below. Does that mean `RootComponentMetadata` can be deleted?

If this is meant for public consumption or some future code we're going to write, we should at least write a test for now.

File: src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs
Comment (halter73) [inline src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs:87]:
```suggestion
            var routingStateProvider = httpContext.RequestServices.GetRequiredService<EndpointRoutingStateProvider>();
            routingStateProvider.SetRouteData(new RouteData(componentType, httpContext.GetRouteData().Values));
```

File: src/Components/Components/src/Routing/Router.cs
Comment (halter73) [inline src/Components/Components/src/Routing/Router.cs:106]: Nit:

```suggestion
        RoutingStateProvider = ServiceProvider.GetService<IRoutingStateProvider>();
```

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs
Comment (javiercn) [reply src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs:43]: Root component metadata is for public consumption so that the endpoint can be identified.

File: src/Components/Components/src/Routing/Router.cs
Comment (halter73) [inline src/Components/Components/src/Routing/Router.cs:51]:
```suggestion
    private IRoutingStateProvider? RoutingStateProvider { get; set; }
```

Right? Why is the build not complaining about this?

And nit: should we also move this up above the normally-`[Inject]`ed properties now?

File: src/Components/Components/src/Routing/Router.cs
Comment (halter73) [inline src/Components/Components/src/Routing/Router.cs:209]: Nit:

```suggestion
        if (RoutingStateProvider?.RouteData is { } routeData)
        {
            Log.NavigatingToComponent(_logger, routeData.PageType, locationPath, _baseUri);

            _renderHandle.Render(Found(routeData));
```

File: src/Components/Components/src/Routing/Router.cs
Comment (BrennanConroy) [reply src/Components/Components/src/Routing/Router.cs:51]: > Why is the build not complaining about this?

https://github.com/dotnet/aspnetcore/blob/4afe7f612d104b43b690e71d83c18a8bc48aae2d/src/Components/Components/src/Routing/Router.cs#L4

File: src/Components/Endpoints/src/DependencyInjection/EndpointRoutingStateProvider.cs
Comment (surayya-MS) [reply src/Components/Endpoints/src/DependencyInjection/EndpointRoutingStateProvider.cs]: Sorry, for some reason github isn't commiting the siggested changed. I applied the change manually.

File: src/Components/Components/src/Routing/Router.cs
Comment (surayya-MS) [reply src/Components/Components/src/Routing/Router.cs]: If we use `[Inject]` then it won't be an optional service.

File: src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs
Comment (surayya-MS) [reply src/Components/Endpoints/src/Builder/RazorComponentEndpointFactory.cs:43]: Done

**Pull request merged**
